Re-introduce `-Winline' (but better), and other sundry improvements

mfwitten opened the pull request:
This series re-introduces the `-Winline' compiler option, but does so in a much more tactful way, avoiding "spammy" behavior. In the course of implementing this series, a number of other changes were made along the way, which relate to inlining, optimization, and build parameterization.

The main gist of this series is captured in the following commits from the series:

    * bff79645 scons: Use `CCFLAGS' instead of `CFLAGS'
    * 8f620409 scons, Travis CI: Set optimization/debugging levels more appropriately
    * d38baff9 scons, config.h, fts: Tidy up debugging assertions
    * 24edd3d3 xxhash: Use `g_slice_alloc()' and `g_slice_free1()'
    * 719ec90b scons, blake2, xxhash: Tidy up inline functions

Some of those are actually merge commits, whose total graph expands to this:

    * bff79645 scons: Use `CCFLAGS' instead of `CFLAGS'
    * 8f620409 scons, Travis CI: Set optimization/debugging levels more appropriately
    |\  
    | * 58487789 Travis CI: Build with `DEBUG=1 CCFLAGS=-g0'
    | * a829faa7 scons: Allow command-line specification of last `CCFLAGS' items
    | * 32708542 scons: Make `-O2' the default optimization level
    |/  
    * d38baff9 scons, config.h, fts: Tidy up debugging assertions
    |\  
    | * 68ab7094 scons: When not `DEBUG=1', define `NDEBUG'
    | * 3ab48a01 lib/fts/fts.c: Use `g_assert()'
    |/  
    * 24edd3d3 xxhash: Use `g_slice_alloc()' and `g_slice_free1()'
    * 719ec90b scons, blake2, xxhash: Tidy up inline functions
    |\  
    | * 60abd103 blake2, xxhash: Bring `BLAKE2_INLINE' and `FORCE_INLINE' in line with `INLINE'
    | * d6c5ec30 scons: Set `-Winline' only when appropriate
    |/  
    * f241bd1b scons: remove spammy -Winline

sahib commented:
> This series re-introduces the `-Winline' compiler option, but does so in a much more tactful way, avoiding "spammy" behavior.

Thanks and sorry for not telling you I dropped ``-Winline``. I originally assumed you left it there by accident.
Maybe a bit background on that: A lot of users get confused when they see warnings when compiling ``rmlint`` and they think something is wrong. So naturally I try to fix them as soon as possible. ``-Winline`` has the disadvantage that those warnings do not tell anything to the user (they're only developer relevant) and might be triggered by using different compiler options (probably common to Gentoo users e.g.). What speaks for keeping ``-Winline`` in all release builds then? Developers can still enable it and I wonder if it should be rather enabled when the developers uses a special variable like ``DEBUG_INLINE`` or something. Would be happy to hear your stance about that.

mfwitten commented on .travis.yml:
```diff
@@ -17,7 +17,7 @@ notifications:
 
 script:
-    - scons VERBOSE=1
+    - scons VERBOSE=1 DEBUG=1 CCFLAGS=-g0
```
Extra Testing
-----------

I agree that it should be as close to the user's setup as possible; however, I always thought of a debug build as being:

* `the user's setup` + `a lot of extra runtime checks` - `optimizations`

For instance, without a debug build, the `g_assert()` calls are removed, which removes testing for programming errors, thereby potentially letting bugs through.

I suppose one could worry about bugs being introduced by the optimizations, and thus one might desire to test an optimized program. I have 2 points on that:

* One solution is to allow for an optimized build with assertions still enabled (that is, decouple optimization and assertions).
* Another solution is to trust that the compiler makers have tested their optimizations well enough, so that one may assume they don't introduce bugs (i.e., just not worry about it).

Personally, I lean towards trusting the compilers, but I can also see the virtue in decoupling optimization and assertions.

Efficiency of Testing
---------------------

However, optimizations are not necessarily a good thing here, as they must be performed every single time the tests are run.

<details>
<summary>To evaluate this, here are some bash functions.</summary>

    #!/bin/bash
    
    msg() { printf "$@" >&3; }
    
    iteration() { msg '(%s of 2) %s\n' "$1" "$2"; }
    
    test()
    (
        set -e
        exec 4>"$1" 3>&1 2>/dev/null 1>&2; shift
        command="scons ${1:+"$1 "}test"
        git clean -fdx
        iteration 1 "$command"
        (
            time \
            {
                eval "$command"
                git clean -fdx
                iteration 2 "$command"
                eval "$command"
            } 2>/dev/null
        ) 2>&4
    )
    
    separate()
    {
        (($#)) &&
            msg '\n%s\n' "=========== $1:" ||
            msg '%s\n' -----------
    }
    
    compare()
    (
        set -e
        exec 3>&1 2>/dev/null 1>&2
    
        name1=${1%%:*}
        if [[ -z $name1 ]]; then
            name1=Test\ 1
        fi
        args1=${1#*:}
    
        name2=${2%%:*}
        if [[ -z $name2 ]]; then
            name2=Test\ 2
        fi
        args2=${2#*:}
    
        separate 'Setup'
        msg "Checking out \"$head\"... "
        git clean -fdx
        git checkout "$head"
        msg 'finished\n'
    
        msg 'Warmup... '
        eval scons "$args1" test
        msg 'finished\n'
        separate 'Running Tests'
        test /tmp/test1 "$args1" 1>&3
        separate
        test /tmp/test2 "$args2" 1>&3
    
        separate "$name1"
        cat /tmp/test1 >&3
        separate "$name2"
        cat /tmp/test2 >&3
    )
</details><br>

The time it takes to build and test a non-debug build can be compared easily to the corresponding time that results from the proposed changes:

    $ head=bff796458872a1781199cd72eaef30f7f891218d
    $ compare Normal: Debug:'DEBUG=1 CCFLAGS=-g0'
    
    =========== Setup:
    Checking out "bff796458872a1781199cd72eaef30f7f891218d"... finished
    Warmup... finished
    
    =========== Running Tests:
    (1 of 2) scons test
    (2 of 2) scons test
    -----------
    (1 of 2) scons DEBUG=1 CCFLAGS=-g0 test
    (2 of 2) scons DEBUG=1 CCFLAGS=-g0 test
    
    =========== Normal:
    
    real    7m30.190s
    user    3m31.650s
    sys     0m46.404s
    
    =========== Debug:
    
    real    7m16.059s
    user    3m11.495s
    sys     0m46.223s

I ran that command twice, and came up with the same results:

* The proposed changes make the testing significantly faster (by about 10 seconds per cycle on my machine).

Because this program is undoubtedly I/O bound, and because the debug build performs many more run-time checks, I suspect that a great deal of the savings come from the compilation phase. A further test bears this out (the `#` comments out the `test` target, thereby running just a build):

    $ compare Normal:'#' Debug:'DEBUG=1 CCFLAGS=-g0 #'
    
    =========== Setup:
    Checking out "bff796458872a1781199cd72eaef30f7f891218d"... finished
    Warmup... finished
    
    =========== Running Tests:
    (1 of 2) scons # test
    (2 of 2) scons # test
    -----------
    (1 of 2) scons DEBUG=1 CCFLAGS=-g0 # test
    (2 of 2) scons DEBUG=1 CCFLAGS=-g0 # test
    
    =========== Normal:
    
    real    1m47.417s
    user    1m37.621s
    sys     0m6.775s
    
    =========== Debug:
    
    real    1m8.520s
    user    0m59.875s
    sys     0m6.445s

Wow! Look at that! Each debug build is around 19 seconds faster on my machine.

Conclusion
----------

Therefore, *currently*, the proposed changes maximize the usefulness of the testing cycle; the result is a wider search for bugs and yet a faster turnaround.

What do you think?

sahib replied:
> For instance, without a debug build, the g_assert() calls are removed, which removes testing for programming errors, thereby potentially letting bugs through.

You have a point there, I forgot about the new ``g_assert()`` relation with ``DEBUG=1``.

> Personally, I lean towards trusting the compilers, but I can also see the virtue in decoupling optimization and assertions.

I would lean to introducing to decoupling of assertions and debug symbols. You never know what compiler/versions of compilers people use to compile ``rmlint``. 

> Therefore, currently, the proposed changes maximize the usefulness of the testing cycle; the 
 result is a wider search for bugs and yet a faster turnaround.
>
> What do you think?

Well, nice... but: We're still talking about the setup on TravisCI here. A service that runs in background and is usually (rightly) totally forgotten until something breaks. It's really one of the last things I'd ever optimize. If it takes a few seconds more to send me a mail notification I wouldn't even notice it. I would not trade it against testing an optimized build.

mfwitten commented on SConstruct:
```diff
@@ -11,7 +11,7 @@ import SCons
 from SCons.Script.SConscript import SConsEnvironment
 
-DEFAULT_OPTIMISATION='s'  # compile with -Os
+DEFAULT_OPTIMISATION='2'  # compile with -O2
```
I agree that this program is likely I/O-bound, and that it makes sense to have chosen the kind of optimization that results in the most noticeable improvement (in this case, *size* optimization rather than *speed* optimization). However, as you know and as you've implied, storage (or bandwidth) is considered very cheap these days, so maybe optimizing for size isn't an overall benefit to the user.

I wish that software were always optimized for speed instead, no matter how small the improvement. This is because I'm never running just one program; I'm never running *just* `rmlint`, and it seems to me that even tiny inefficiencies *accumulate* to produce noticeable degredation&mdash;no *one* program in particular is to blame, but the *group* as a whole *is* to blame.

<details>
<summary>These <code>bash</code> functions provide basic benchmarking.</summary>

    #!/bin/bash
    
    ITERATIONS=10
    
    msg() { printf "$@" >&3; }
    
    clean_up()
    {
        find /tmp -type d -not -path /tmp -prune -o \
            -type f \( -name out\*.json -o -name replay\*.json \) \
                -exec rm {} +
    
        find ./ -type d -not -path ./ -prune -o \
            -type f -name rmlint.\* \
                -exec rm {} +
    
        rm -f .rmlint.json
    }
    
    run_tests()
    {
        RM_TS_PRINT_CMD=1 RM_TS_PEDANTIC=0 nosetests -s -v -a '!slow'
    }
    
    warmup()
    {
        local compile="scons${1:+" $1"}"
        msg '%s\n' "$compile"
        msg 'Warmup... '
        git clean -fdx
        eval "$compile"
        run_tests
        clean_up
        msg 'finished\n'
    }
    
    benchmark()
    {
        msg 'Testing... 0'
        {
            local i
            time for ((i=1; i<=ITERATIONS; ++i)); do
                run_tests
                msg '\b%s' "$i"
            done 1>/dev/null 2>&1
        } 2>&1
        clean_up >/dev/null
        msg '\b\bfinished\n'
    }
    
    average_time()
    {
        {
            printf '%s\n' 'scale=3; total=0'
            while read clock time; do
                if [[ $clock =~ user|sys ]]; then
                    if [[ $time =~ ([^m]+)m([^s]+)s ]]; then
                        printf '%s\n' "total += ${BASH_REMATCH[1]}*60 + ${BASH_REMATCH[2]}"
                    else
                        return 1
                    fi
                fi
            done
            printf '%s\n' "total / $ITERATIONS"
        } | bc
    }
    
    test()
    (
        set -e
        exec 5>"$1" 4<"$1" 3>&1 2>/dev/null 1>&2
        shift
    
        warmup "$1"
        benchmark >&5
    
        msg '%s: %s seconds\n' 'Average time' "$(average_time <&4)"
    )
    
    separate()
    {
        (($#)) &&
            msg '\n%s\n' "=========== $1:" ||
            msg '%s\n' -----------
    }
    
    compare()
    (
        set -e
        exec 3>&1 2>/dev/null 1>&2
    
        if (($# > 2)); then
            msg '%s\n' 'Usage: compare [args1 [args2]]\n'
            return 1
        fi
    
        name1=${1%%:*}
        if [[ -z $name1 ]]; then
            name1=Test\ 1
        fi
        args1=${1#*:}
    
        name2=${2%%:*}
        if [[ -z $name2 ]]; then
            name2=Test\ 2
        fi
        args2=${2#*:}
    
        separate 'Setup'
        msg "Checking out \"$head\"... "
        git clean -fdx
        git checkout --detach "$head"
        msg 'finished\n'
    
        separate 'Running Tests'
        test /tmp/test1 "$args1" >&3
        if (($# != 2)); then
            separate "$name1"
            cat /tmp/test1 >&3
        else
            separate
            test /tmp/test2 "$args2" >&3
    
            separate "$name1"
            cat /tmp/test1 >&3
            separate "$name2"
            cat /tmp/test2 >&3
        fi
    )
</details><br>

The following commands build `rmlint` with the provided build-time arguments, and then calculate the average running time of the test suite across 10 iterations for that particular build:

    $ head=bff796458872a1781199cd72eaef30f7f891218d
    $ compare Size:O=s Speed:O=2
    
    =========== Setup:
    Checking out "bff796458872a1781199cd72eaef30f7f891218d"... finished
    
    =========== Running Tests:
    scons O=s
    Warmup... finished
    Testing... finished
    Average time: 77.231 seconds
    -----------
    scons O=2
    Warmup... finished
    Testing... finished
    Average time: 76.305 seconds
    
    =========== Size:
    
    real    28m14.593s
    user    9m33.759s
    sys     3m18.552s
    
    =========== Speed:
    
    real    27m53.960s
    user    9m25.998s
    sys     3m17.052s

So, when optimized for speed, the test suite is 0.926 seconds faster. That's not much for a human, but for a computer, that's a damn long time! I've got another little patch series in the works, which involves some inlining optimization that resulted from refactoring, and it might shave off a little more time:

    $ head=branch-name
    $ compare Speed:O=2
    
    =========== Setup:
    Checking out "branch-name"... finished
    
    =========== Running Tests:
    scons O=2
    Warmup... finished
    Testing... finished
    Average time: 75.920 seconds
    
    =========== Speed:
    
    real    27m50.822s
    user    9m25.630s
    sys     3m13.571s

That's a difference of 0.385 seconds.  
I wonder how many other fractions of a second could be shaved away.

sahib replied:
> I wish that software were always optimized for speed instead, no matter how small the improvement. This is because I'm never running just one program; I'm never running just rmlint, and it seems to me that even tiny inefficiencies accumulate to produce noticeable degredation—no one program in particular is to blame, but the group as a whole is to blame.

True, but that's only one metric to optimize. Each optimization is a trade-off. Sometimes you just trade CPU against memory. Sometimes you treat maintainability or being able to easily debug something against performance. Not claiming this is the case here, just saying that general goal *"more speed!"* shouldn't be always followed under all circumstances.

> These bash functions provide basic benchmarking.

Thank you for doing those. 

> The following commands build rmlint with the provided build-time arguments, and then calculate the average running time of the test suite across 10 iterations for that particular build:

Not sure if the test-suite is a very good benchmark for it. There's a questionable amount of time that goes into Python and each ``rmlint`` call usually only gets called on a tiny amount of data.
For quick and dirty benchmark I usually just copy some bigger file tree (parts of ``/usr``) on tmpfs and let it run there. What you want to test are longer running ``rmlint`` processes.

> That's a difference of 0.385 seconds.
> I wonder how many other fractions of a second could be shaved away.

Again, it's mostly I/O-bound and I'm not sure if your script really benchmarks the important bits of ``rmlint`` (i.e. the incremental hashing). Due to the high amount of calls to ``rmlint`` in the test suite the result might just be due to the faster startup (maybe faster command line processing and formatting). Not saying that the results themselves are not valid, but just saying that this might be a flaw in the benchmark.

But apart from this nitpicking: I guess you convinced me enough already to switch to ``-O2``. :smirk:

mfwitten replied:
Good points.

I'll note that *currently*, normal usage of `rmlint` might also involve a lot of startup costs; for instance, `rmlint --dedupe` may be called very many times indeed.

mfwitten commented:
# Changes

* I've removed `-Winline`.

    It's not possible to cator to both end-users and developers, and so it's better to cator to end-users; an interested developer can use one of the relevant environment variables or command-line variables (`CFLAGS`, or `CCFLAGS`, etc.) to specify `-Winline`.

* The `DEBUG=1` option no longer tells the compiler to emit debugging symbols; instead, one may use `SYMBOLS=1`. For convience, one may also use `GDB=1` to mean `DEBUG=1 SYMBOLS=1`.

    Of course, more customized control  is available with the usual compiler variables.

* The `CCFLAGS` command-line variable is now parsed with `shlex.split()`.

# Review

I've found it helpful to review the patches with the following command line:

    $ git log --reverse -p --topo-order origin/develop..

Here is the new graph:

    * a451f17e [HEAD] scons: Make dict access more "Pythonic"
    * d58dd449 scons: Tidy up `CCFLAGS'
    |\  
    | * 96d80155 scons: Use `CCFLAGS' instead of `CFLAGS'
    | * 7be8b372 scons: Allow command-line specification of last `CCFLAGS' items
    |/  
    * 8c58eeee scons, Travis CI: Set optimization/debugging levels more appropriately
    |\  
    | * a2887051 Travis CI: Build with `DEBUG=1'
    | * 7c9d990a scons: Introduce `GDB=1'
    | * bf1d7033 scons: Introduce `SYMBOLS=1'
    | * ffd665a2 scons: Make `-O2' the default optimization level
    | * da84c3ab scons: Remove `DEFAULT_OPTIMISATION'
    |/  
    * 3ffdbc7a scons, config.h, fts: Tidy up debugging assertions
    |\  
    | * 7b776b8d scons: When not `DEBUG=1', define `NDEBUG'
    | * 195965bb lib/fts/fts.c: Use `g_assert()'
    |/  
    * 2c53c099 xxhash: Optimize and modernize allocation/deallocation
    |\  
    | * 067041fd xxhash: `s/malloc/new/' and `s/free/delete/'
    | * 6069b1a6 xxhash: Use `g_slice_alloc()' and `g_slice_free1()'
    |/  
    * ca6f6c2f blake2, xxhash: Tidy up inline functions
    |\  
    | * be03afcb blake2, xxhash: Use `INLINE' directly
    | * 5330cf41 blake2, xxhash: Bring `BLAKE2_INLINE' and `FORCE_INLINE' in line with `INLINE'
    |/  
    * 24318da8 [develop]

sahib commented on .travis.yml:
```diff
@@ -17,7 +17,7 @@ notifications:
 
 script:
-    - scons VERBOSE=1
+    - scons VERBOSE=1 DEBUG=1
```
As discussed earlier, I'd love to have TravisCI still run with an optimized build (which ``DEBUG=1`` sets to ``-O0``).
Suggestion: Make DEBUG=1 still honor the ``O=2`` switch when it's set explicitly.

sahib commented on SConstruct:
```diff
@@ -713,20 +711,34 @@ conf.check_sysmacro_h()
 
-# compiler optimisation and debug symbols:
-cc_O_option = '-O'
+if ARGUMENTS.get('GDB') == '1':
```
``docs/developers.rst`` needs to be updated with the ``GDB`` and ``SYMBOLS`` variables.

sahib commented:
Beside the TravisCI discussion this PR looks good to me already. :+1:

mfwitten commented:
Here is the overall diff since the last version:

    diff --git a/.travis.yml b/.travis.yml
    index 336591e8..99b462b5 100644
    --- a/.travis.yml
    +++ b/.travis.yml
    @@ -17,7 +17,7 @@ notifications:
         - thomas_d_j@yahoo.com.au
     
     script:
    -    - scons VERBOSE=1 DEBUG=1
    +    - scons VERBOSE=1 DEBUG=1 O=release
         - scons config
         - export RM_TS_PRINT_CMD=1
         - export RM_TS_PEDANTIC=0
    diff --git a/SConstruct b/SConstruct
    index 835d2284..ec41404f 100755
    --- a/SConstruct
    +++ b/SConstruct
    @@ -715,15 +715,23 @@ if ARGUMENTS.get('GDB') == '1':
         ARGUMENTS['DEBUG'] = '1'
         ARGUMENTS['SYMBOLS'] = '1'
     
    -# optimisation modes
    +O_DEBUG   = 'g' # The optimisation level for a debug   build
    +O_RELEASE = '2' # The optimisation level for a release build
    +
    +# build modes
     if ARGUMENTS.get('DEBUG') == "1":
         print("Compiling in debug mode")
         conf.env.Append(CCFLAGS=['-DRM_DEBUG', '-fno-inline'])
    -    O_value = ARGUMENTS.get('O', '0')
    +    O_value = ARGUMENTS.get('O', O_DEBUG)
     else:
         conf.env.Append(CCFLAGS=['-DG_DISABLE_ASSERT', '-DNDEBUG'])
         conf.env.Append(LINKFLAGS=['-s'])
    -    O_value = ARGUMENTS.get('O', '2')
    +    O_value = ARGUMENTS.get('O', O_RELEASE)
    +
    +if O_value == 'debug':
    +    O_value = O_DEBUG
    +elif O_value == 'release':
    +    O_value = O_RELEASE
     
     cc_O_option = '-O' + O_value
     
    diff --git a/docs/developers.rst b/docs/developers.rst
    index 98e43454..495f4820 100644
    --- a/docs/developers.rst
    +++ b/docs/developers.rst
    @@ -100,16 +100,70 @@ Environment Variables
     Variables
     ~~~~~~~~~
     
    +:O=<level>:
    +
    +    Set the optimization level.
    +
    +    Valid levels are currently those that may be passed with the GCC/Clang
    +    option ``-O``; these include ``0``, ``1``, ``2``, ``3``, ``s``,
    +    ``fast``, ``g``, etc., depending on the compiler version.
    +
    +    In addition, the level may be ``debug`` or ``release``, which indicates
    +    that the optimization level should be whatever the build system currently
    +    defines to be the default for the associated build mode.
    +
    -:DEBUG:
    -
    -    Enable debugging symbols for ``rmlint``. This should always be enabled during
    -    development. Backtraces wouldn't be useful elsewhise.
    -
    +:DEBUG=1:
    +
    +    Enable a debugging build.
    +
    +    This turns on extra tests; in particular, it turns on run-time
    +    assertions. By default, a debug build excludes optimizations that may
    +    hinder debugging, but this may be overridden with the ``O`` variable,
    +    as usual.
    +
    +    Note that setting ``DEBUG=1`` does not enable the production of
    +    debugger symbols; to enable those, use ``SYMBOLS=1`` or ``GDB=1``.
    +
    +    This should always be enabled during development.
    +
    +:SYMBOLS=1:
    +
    +    Enable debugger symbols.
    +
    +    This option instructs the compiler to collect information that will
    +    help tools such as ``gdb`` present human-readable identifiers for
    +    a program's functions and variables, etc. Note, though, that this
    +    information becomes obscured by optimizations, so make sure to set
    +    the optimization level appropriately.
    +
    +:GDB=1:
    +
    +    Enable options that help a debugger (such as ``gdb``).
    +
    +    This option is equivalent to ``DEBUG=1 SYMBOLS=1``.
    +
    -:VERBOSE:
    +:VERBOSE=1:
     
         Print the exact compiler and linker commands. Useful for troubleshooting
         build errors.
     
    +:CCFLAGS=<command line options>:
    +
    +    Set the last compiler options.
    +
    +    Internally, the build system maintains in ``CCFLAGS`` the list of
    +    options that are supplied to the compiler; this list is composed by
    +    combining the relevant environment variables (such as ``CFLAGS``)
    +    along with the choices made by other build-time configurations.
    +
    +    This command-line variable makes it possible to override an option in
    +    this list by supplying customized command-line options to be appended.
    +    For example: ``GDB=1 CCFLAGS=-g1``.
    +
    +    The string that is supplied as the value for this variable is parsed as
    +    per a POSIX shell command line, and so it may include shell quoting if
    +    necessary.
    +
     Arguments
     ~~~~~~~~~
     
Here is the overall graph (with some new commits expanded):

    * 7a7d1356 [HEAD] scons: Make dict access more "Pythonic"
    *   49be4c59 scons: Tidy up `CCFLAGS'
    |\  
    | * bceecc0b scons: Use `CCFLAGS' instead of `CFLAGS'
    | * e1416130 scons: Allow command-line specification of last `CCFLAGS' items
    |/  
    *   0e68967a scons, Travis CI: Set optimization/debugging levels more appropriately
    |\  
    | * fcbafaf1 Travis CI: Build with `DEBUG=1 O=release'
    | * 8507d66b scons: Introduce `SYMBOLS=1' and `GDB=1'
    | * 98f88e5f scons: Make `-Og' the default optimization level for a debug build
    | | 
    | |    On 22 March 2013, `gcc` version 4.8.0 was released, with the following
    | |    new optimization level:
    | |    
    | |        -Og
    | |    
    | |    The change log is located here:
    | |    
    | |        https://gcc.gnu.org/gcc-4.8/changes.html
    | |    
    | |    It explains this option as follows:
    | |    
    | |        A new  general optimization  level, -Og,  has been  introduced. It
    | |        addresses the need  for fast compilation and  a superior debugging
    | |        experience  while   providing  a  reasonable  level   of  run-time
    | |        performance. Overall  experience for development should  be better
    | |        than the default optimization level -O0.
    | |    
    | |    It's documented here for GCC version 7.3.1:
    | |    
    | |        info '(gcc)Debugging Options'
    | |        info '(gcc)Optimize Options'
    | |    
    | |    It states:
    | |    
    | |         If you  are not  using some  other optimization  option, consider
    | |         using '-Og'  (*note Optimize Options::)  with '-g'. With  no '-O'
    | |         option  at all,  some  compiler passes  that collect  information
    | |         useful for debugging do not run  at all, so that '-Og' may result
    | |         in a better debugging experience.
    | |    
    | |         ...
    | |    
    | |        '-Og'
    | |             Optimize  debugging experience.  '-Og' enables  optimizations
    | |             that  do   not  interfere   with  debugging.  It   should  be
    | |             the   optimization  level   of   choice   for  the   standard
    | |             edit-compile-debug  cycle,  offering  a reasonable  level  of
    | |             optimization while  maintaining fast  compilation and  a good
    | |             debugging experience.
    | |
    | * d78ed3d2 scons: Make `-O2' the default optimization level for a release build
    | * a8184078 scons: Abstract out optimization default for release/debug mode
    | |
    | |    scons: Abstract out optimization default for release/debug mode
    | |    
    | |    There are now 2 "constants":
    | |    
    | |        O_DEBUG
    | |        O_RELEASE
    | |    
    | |    Each  represents the  default optimization  level that  should be
    | |    used for its respective build mode,  and can be referenced on the
    | |    command line:
    | |    
    | |        O=debug
    | |        O=release
    | |    
    | |    Thus,  if   it  were  ever   desirable  to  change   the  default
    | |    optimization level  for one of  these modes, then one  would need
    | |    only alter the relevant constant.
    | |
    | * 2b0882a5 scons: Remove `DEFAULT_OPTIMISATION'
    |/  
    *   2d94d0a3 scons, config.h, fts: Tidy up debugging assertions
    |\  
    | * 750957dd scons: NDEBUG <-> !RM_DEBUG
    | * bb3998d1 lib/fts/fts.c: Use `g_assert()'
    |/  
    *   ca05f1cb xxhash: Optimize and modernize allocation/deallocation
    |\  
    | * 41a8965b xxhash: `s/malloc/new/' and `s/free/delete/'
    | * 0c75a0f8 xxhash: Use `g_slice_alloc()' and `g_slice_free1()'
    |/  
    *   8e76e7d5 blake2, xxhash: Tidy up inline functions
    |\  
    | * 2b19e1cf blake2, xxhash: Use `INLINE' directly
    | * b21e7f2f blake2, xxhash: Bring `BLAKE2_INLINE' and `FORCE_INLINE' in line with `INLINE'
    |/  
    * 24318da8 [develop]

sahib commented:
Looks good to me now. Thanks for the changes and I'm really amazed by the level of detail you put into your commits and responses. Good work.

Aside: Do you plan any other bigger series (besides the mentioned inlining)? I'm asking because I'm thinking of doing a release of ``rmlint`` in the next week(s), so that the bug fixes until now reaches the users. I wondered if you have anything cooking up that deserves to get in the release.

mfwitten commented:
OK. Make the release when you see fit.
